File: JanperCab/5-JanperCab/ClientApp/src/app/_models/duraform-order/DuraformFormDto.ts
import { DuraformMiscDto } from './../duraform-misc/DuraformMiscDto';
import { DuraformFileDto } from './../application-file/DuraformFileDto';
import { DuraformDrawerDto } from './../duraform-component/DuraformDrawerDto';
import { DuraformEndPanelDto } from './../duraform-component/DuraformEndPanelDto';
import { DuraformPantryDoorDto } from './../duraform-component/DuraformPantryDoorDto';
import { DuraformOrderTypeKey } from './../../_enums/DuraformOrderTypeKey';
import { DuraformDoorDto } from './../duraform-component/DuraformDoorDto';
import { Type } from 'class-transformer';
import { DuraformComponentDto } from '../duraform-component/DuraformComponentDto';
import * as _ from 'lodash';

export abstract class DuraformFormDto {
  id: string;
  orderType: DuraformOrderTypeKey;
  customerOrderNumber: string;
  duraformDesignId: number;
  duraformSerieId: number;
  isRoutingOnly: boolean;
  duraformWrapTypeId: number;
  duraformWrapColorId: number;
  duraformEdgeProfileId: number;
  hingeHoleTypeId: number;
  duraformArchId: number;
  createdDate: Date;
  lastUpdated: Date;

  distributorId: number;
  cabinetMakerId: number;

  invoiceTo: string;
  invoiceAddress: string;
  invoiceSuburb: string;
  invoiceState: string;
  invoicePostcode: string;
  deliveryTo: string;
  deliveryAddress: string;
  deliverySuburb: string;
  deliveryState: string;
  deliveryPostcode: string;
  deliveryNote: string;

  notEditable: boolean;
  totalPrice: number;

  @Type(() => DuraformComponentDto, {
    keepDiscriminatorProperty: true,
    discriminator: {
      property: '$type',
      subTypes: [
        {
          value: DuraformDoorDto,
          name:
            '_3_Application.Dtos.DuraformComponent.DuraformDoorDto, 3-Application',
        },
        {
          value: DuraformPantryDoorDto,
          name:
            '_3_Application.Dtos.DuraformComponent.DuraformPantryDoorDto, 3-Application',
        },
        {
          value: DuraformEndPanelDto,
          name:
            '_3_Application.Dtos.DuraformComponent.DuraformEndPanelDto, 3-Application',
        },
        {
          value: DuraformDrawerDto,
          name:
            '_3_Application.Dtos.DuraformComponent.DuraformDrawerDto, 3-Application',
        },
      ],
    },
  })
  duraformComponents: DuraformComponentDto[] = [];
  duraformMiscs: DuraformMiscDto[] = [];
  duraformFiles: DuraformFileDto[] = [];

  get discriminator(): string {
    return `${this.orderType}`;
  }

  get duraformDoors(): DuraformDoorDto[] {
    const doors = this.duraformComponents.filter(
      (x) => x instanceof DuraformDoorDto
    );

    return _.orderBy(doors as DuraformDoorDto[], ['sortNumber'], ['asc']);
  }

  get pantryDoors(): DuraformPantryDoorDto[] {
    const pantryDoors = this.duraformComponents.filter(
      (x) => x instanceof DuraformPantryDoorDto
    );

    return _.orderBy(
      pantryDoors as DuraformPantryDoorDto[],
      ['sortNumber'],
      ['asc']
    );
  }

  get endPanels(): DuraformEndPanelDto[] {
    const endPanels = this.duraformComponents.filter(
      (x) => x instanceof DuraformEndPanelDto
    );

    return _.orderBy(
      endPanels as DuraformEndPanelDto[],
      ['sortNumber'],
      ['asc']
    );
  }

  get duraformDrawers(): DuraformDrawerDto[] {
    const drawers = this.duraformComponents.filter(
      (x) => x instanceof DuraformDrawerDto
    );

    return _.orderBy(drawers as DuraformDrawerDto[], ['sortNumber'], ['asc']);
  }

  constructor() {}
}

File: JanperCab/3-Application/Dtos/DuraformForm/DuraformFormDto.cs
using _1_Domain;
using _3_Application.Dtos.DuraformComponent;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel.DataAnnotations;

namespace _3_Application.Dtos.DuraformForm
{
    public abstract class DuraformFormDto
    {
        public Guid Id { get; set; }

        public DuraformOrderType OrderType { get; set; }

        [Required]
        public string CustomerOrderNumber { get; set; }

        public int DuraformDesignId { get; set; }

        public _1_Domain.DuraformSerie.DuraformSerieKey DuraformSerieId { get; set; }

        public bool IsRoutingOnly { get; set; }

        public int? DuraformWrapTypeId { get; set; }

        public int? DuraformWrapColorId { get; set; }

        public int DuraformEdgeProfileId { get; set; }

        public int? HingeHoleTypeId { get; set; }

        public int? DuraformArchId { get; set; }

        public DateTime CreatedDate { get; set; }

        public DateTime? LastUpdated { get; set; }


        public int DistributorId { get; set; }

        public int CabinetMakerId { get; set; }


        public string InvoiceTo { get; set; }

        public string InvoiceAddress { get; set; }

        public string InvoiceSuburb { get; set; }

        public string InvoiceState { get; set; }

        public string InvoicePostcode { get; set; }

        public string DeliveryTo { get; set; }

        public string DeliveryAddress { get; set; }

        public string DeliverySuburb { get; set; }

        public string DeliveryState { get; set; }

        public string DeliveryPostcode { get; set; }

        public string DeliveryNote { get; set; }

        public bool NotEditable { get; set; }

        public decimal? TotalPrice { get; set; }


        public ICollection<DuraformComponentDto> DuraformComponents { get; set; }

        protected DuraformFormDto()
        {
            DuraformComponents = new Collection<DuraformComponentDto>();
        }
    }
}